### ISSUE TITLE
Run services as non-privileged users

### ISSUE THREAD

**Issue by stephenR** [GoogleCloudPlatform/kubernetes#7961]
At the moment most services in kubernetes run as root. This includes (from glancing at ps aux on a minion):
* etcd
* kube2sky
* td-agent
* skydns
* elasticsearch
* heapster
It's best practice to use non-root user accounts for services that don't need them. E.g. elasticsearch had remote code execution vulnerabilities in the past.
Even though docker's root has a capability bounding set applied so it doesn't translate to a real root on the host, the attack surface for privilege escalation attacks is still increased.

**Comment by erictune** [comment]
@thockin I think you have given this issue some thought.  What did you want to do about low numbered ports? Run stuff as non root and make it easy to get CAP_NET_BIND?  

@pweil- @pmorie what does OpenShift do about CAP_NET_BIND and default Capabilities for SecurityContexts?

**Comment by thockin-cc** [comment]
More or less, yeah. I think we have a responsibility as thought leaders to
push this direction. There's no reason to do a allow low-ports for non-priv
pods (namespaces) or for privileged net=host pods.

**Comment by erictune** [comment]
I don't understand "there's no reason to do a allow low-ports for non-priv
pods (namespaces) or for privileged net=host pods."

**Comment by erictune** [comment]
I agree we don't want CAP_NET_BIND for net-host pods.

I thought we would want to allow CAP_NET_BIND for all other containers, but your words seem to contradict that.

**Comment by thockin** [comment]
Sorry, typing on a phone. I see no reason to disallow low ports on normal
containers or privileged containers with net=host. That covers the vast
majority of containers. I don't know if that alone is sufficient to get
people off UID 0, but it's certainly part of the solution.

Tim

**Comment by timstclair** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/38541

**Comment by kargakis** [comment]
@kubernetes/sig-auth-misc

**Comment by liggitt** [comment]
@tallclair is this mostly addressed by the PSP bounding you did for the addons we run?

**Comment by tallclair** [comment]
No. I made minimal changes to the deployments when adding the per-service profiles. Most of these services definitely do not need to run as root. A handful of them require root just for writing their log files, but there's no reason the log files need to be owned by root...

**Comment by mourya007** [comment]
@liggitt @tallclair Is this bug still exist ?

**Comment by tallclair** [comment]
Yep, and help is still wanted :)

**Comment by mourya007** [comment]
@timstclair Thanks, Please Let me know how can i give my help in this issues.

**Comment by tallclair** [comment]
The hardest part of this is going to be handling the log files. If we can ensure those (and any other host mounted files) have the right permissions, then setting RunAsUser should be pretty straightforward (plus adequate testing).

**Comment by neolit123** [comment]
/remove-sig cluster-lifecycle
none of the listed components are owned by the sig, yet we do aknowledge the issue at kubeadm level:
https://github.com/kubernetes/kubeadm/issues/1367

**Comment by liggitt** [comment]
sig-auth does not own configuration of cluster addons. Several of the mentioned addons are no longer run at all. This unowned umbrella issue is clearly not making progress. I would suggest:
1. identify which components are being deployed by kube-up or kubeadm
2. create individual issues tracking running them with reduced privileges
3. close this issue

/assign @vinayakankugoyal

**Comment by pacoxu** [comment]
As far as I know, take coredns as an example, it doesn't support non-root as it needs to bind to 53 port. 
If changing to non-root, the port needs to be bigger which will cause backward issues.

Per https://github.com/coredns/coredns/pull/3923#issuecomment-637969242

**Comment by vinayakankugoyal** [comment]
@pacoxu - coredns can be run as non-root and bind to port 53 if we apply cap_net_bind_service file capabilities to it at build time. This is what we did for kube-apiserver. See these PRs:
1. https://github.com/kubernetes/kubernetes/pull/99145
2. https://github.com/kubernetes/kubernetes/pull/96134

@liggitt - Looks like the issue can't be assigned to me. 

I am currently working on a KEP for running the control-plane in kubeadm as non-root here: https://github.com/kubernetes/enhancements/pull/2569

**Comment by vinayakankugoyal** [comment]
for coredns we could do something like: https://github.com/coredns/coredns/pull/4528

**Comment by pacoxu** [comment]
@vinayakankugoyal 
Thanks for the information. 

You can comment with `/assign` even not a member. (or open issue https://github.com/kubernetes/org/issues to join the org 🎉)

**Comment by vinayakankugoyal** [comment]
/assign 
@pacoxu - I don't have an outside of google member to sponsor me yet. I am hoping to join the org one day!

**Comment by LyKos4** [comment]
Is there a solution to run all pods that run in kube-system namespace (needed for a kubernetes cluster) as non root? 
Are there security danged when running them as root? Even if they aren't accessible with bash/sh?

**Comment by thockin** [comment]
This question is pretty broad to answer with a yes or no. Different installations have different pods in there. In theory it should be possible, but it depends on exactly what those pods are and what they're doing. Many of them interface with other aspects of the machine, and need some sort of elevated privilege. 

That doesn't mean that everything that runs as root today needs to run as root. There are certainly things in there that could run as non-root, but nobody's made the effort to do that. 

As to security dangers, it's always better to run non-root than root, but we do our best to address security issues as they come up. So it's not like you're sitting on a known bomb.

**Comment by LyKos4** [comment]
Let me make my questions more specific.
For the basic pods running for a kubernetes cluster to be functional:  coredns, etcd, kube-apiserver, kube-controller, kube-proxy, kube-scheduler, are there any that it is known that need to run as root? 
For rest, are there suggested ways to make run as user with less privileges?

**Comment by lmktfy** [comment]
Is this now feasible?
